apm-agent-plugins/apm-scala-concurrent-plugin/src/main/java/co/elastic/apm/agent/scala/concurrent/FutureInstrumentation.java

```diff
@@ -0,0 +1,157 @@
+            final TraceContextHolder<?> active = getActive();
+            if (active != null && runnable != null && tracer != null) {
+                // Do no discard branches leading to async operations so not to break span references
+                active.setDiscard(false);
+                runnable = active.withActive(runnable);
+            }
+        }
+
+    }
+
+    public static class BatchedExecutionContextInstrumentation extends FutureInstrumentation {
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return hasSuperType(named("scala.concurrent.BatchingExecutor"));
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher(){
+            return named("submitForExecution").and(returns(void.class)).and(takesArguments(Runnable.class))
+                .or(named("submitAsyncBatched").and(returns(void.class)).and(takesArguments(Runnable.class)))
+                .or(named("submitSyncBatched").and(returns(void.class)).and(takesArguments(Runnable.class)));
+        }
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        public static void onExecute(@Advice.Argument(value = 0, readOnly = false) @Nullable Runnable runnable) {
+            final TraceContextHolder<?> active = getActive();
+            if (active != null && runnable != null && tracer != null) {
+                // Do no discard branches leading to async operations so not to break span references
+                active.setDiscard(false);
+                runnable = active.withActive(runnable);
+            }
+        }
+
+    }
+
+    public static class ConstructorInstrumentation extends FutureInstrumentation {
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return hasSuperType(named("scala.concurrent.impl.Promise$Transformation"));
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return isConstructor();
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class)
+        public static void onExit(@Advice.This Promise<?> thiz) {
+            final TraceContextHolder<?> active = getActive();
+            if (active != null) {
+                promisesToContext.put(thiz, active);
+            }
+        }
+
+    }
+
+    public static class RunInstrumentation extends FutureInstrumentation {
+
+        @Override
+        public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+            return hasSuperType(named("scala.concurrent.impl.Promise$Transformation"));
+        }
+
+        @Override
+        public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+            return named("run").and(returns(void.class));
+        }
+
+        @VisibleForAdvice
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        public static void onEnter(@Advice.This Promise<?> thiz) {
```

Comment:
To properly deactivate on exit with considering wrapped promises, add `@Advice.Local("context") TraceContextHolder<?> context` to the parameters.

Reply:
Maybe the non-deactivated context has something to do with the MDC issue.

Reply:
Im not using the matchers on the Promise directly anymore. I'm only using matchers on the ExecutionContext directly using the wrapped Runnables:
```
co.elastic.apm.agent.scala.concurrent.FutureInstrumentation$ExecutionContextInstrumentation
co.elastic.apm.agent.scala.concurrent.FutureInstrumentation$BatchedExecutionContextInstrumentation
#co.elastic.apm.agent.scala.concurrent.FutureInstrumentation$ConstructorInstrumentation
#co.elastic.apm.agent.scala.concurrent.FutureInstrumentation$RunInstrumentation
```

Reply:
ah, got it